---
title: "Covid project"
author: "Kristian Riis Hansen"
date: '2022-09-06'
output: html_document
---
Libraries:
```{r}
library(dplyr)
library(tidyverse)
library(ggplot2)
library(lubridate)

```

working directory
```{r}
setwd("C:/Users/Krist/Desktop/PiB/Variants and mutation rate in SARS-cov2")
getwd()

```


Loading data: 
```{r}


tidy_300K <- readRDS("metadata_snpeff_tidy_300K_downsampled.RDS")
mutations_tidy_300K <- readRDS("mutations_snpeff_annotated_tidy_300K_downsampled.rds")

```



#Are virus mutation rates correlated with location/different pangolin lineages?
#GC content

#Are there single mutations that indicate increased spread in populations. (more virulent or something)


```{r}

count(tidy_300K$country)
?count

table(tidy_300K$country)
countries_over_1k <- tidy_300K %>% group_by(country) %>% mutate(Count=n())

countries_over_1k

```



#new dataframe with count over 1k.
```{r}


filterdf <- countries_over_1k %>% filter(Count >1000)

```



#Shows countries with over 1000 samples
```{r}
table(filterdf$country)

View(filterdf)



small_filt <- filterdf %>% select (country,id, `Virus name`, count_N, count_S) %>%
  mutate(S_over_N = count_S/count_N)

```



  
#Removing NaN rows:

```{r}
small_filt <- na.omit(small_filt)

small_filt <- small_filt[is.infinite(rowSums(small_filt)),]


```



#There are still rows with "inf"

```{r}

small_filt %>% 
  ggplot(aes(country, S_over_N)) +
  geom_boxplot() +
  geom_jitter(width = 0.15) +
  theme(axis.text.x = element_text(angle = 45))


?geom_boxplot()


small_filt %>% 
  ggplot(aes(country, S_over_N)) + 
  geom_boxplot() + 
  geom_jitter(width = 0.2) + 
  theme(axis.text.x = element_text(angle = 45))

```








Merging gene coloumn from tidy_300k to mutation dataset:


we want to detect the selective sweep
we found a method for that in one of the articles - so we count the mutations for each positions and we got that around position 22.500 there are significantly more datapoints so we want to analyse that specific region and check the certain gene which might be responsible for the sweep.

```{r}
left <- tidy_300K

right <- mutations_tidy_300K %>% select(1,5)

merged_df <- merge(left, right)

```

```{r}
left_join(left,right, by =c("id"="id", "position"="position"))

```




We want to merge the datasets so we have dates, country, GC-content.



```{r}


test_mutation_df <- mutations_tidy_300K[,]
  
test_tidy_df <- (tidy_300K[,c(1,10,13,31)])

test_mutation_df <- test_mutation_df %>% group_by(id)


test_merged_df <- left_join(test_tidy_df, test_mutation_df, by =c("id"="id"))


test_merged_df <- test_merged_df %>% group_by(position) %>% 
  mutate(count = n())


```




We now look at a specific country and gene and graph some things.

```{r}

dk_merged <- filter(test_merged_df, country == "Denmark")


fra_merged <- filter(test_merged_df, country == "France")

```

Converting date column into numeric. Works for denmark because it has days included. but the ones in france that already have month only, will get NA.

```{r}

Date = format(as.Date(dk_merged$`Collection date`), "%Y - %m")


Date = format(as.Date(fra_merged$`Collection date`), "%Y-%m")


#Trying using substring
fra_merged <- mutate(fra_merged, Date = format(substring(fra_merged$`Collection date`,0,7)))
#Works like a charm.

dk_merged <- dk_merged %>% mutate(Date = Date)


```




Looking at GC-content over time in Denmark.

```{r}

ggplot(data=dk_merged, aes(x=Date, y=dk_merged$"GC-Content")) + 
         geom_boxplot() +
        theme(axis.text.x=element_text(angle=90)) + ylab("GC content")
      



```
GC-content over time in France.

Because france has samples that are labeled on only the years, we will filter these out.

Using grepl to remove the "years only" rows.
```{r}

remove <- c("2020,2021,2022")

fra_merged <- fra_merged %>% grepl(remove, fra_merged$Date)






```


Plotting GC content for france over time.
```{r}

ggplot(data=fra_merged, aes(x=Date, y=fra_merged$"GC-Content")) + 
         geom_boxplot() +
        theme(axis.text.x=element_text(angle=90)) + ylab("GC content")
      



```

Filter danish dataset for only one year and spike gene to see mutation positions.


```{r}

dk_2020 <- dk_merged %>% filter(str_detect(Date, "^2020"))




dk_2020 %>% group_by(position) %>%
  mutate(count = n()) %>%
  filter(gene == "S") %>%
  ggplot(dk_2020, mapping = aes(x=position, y=count)) + geom_point() +
  ggtitle("Spike gene mutations during 2020 in Denmark")



```

Same for france:

```{r}

fra_merged %>% filter(str_detect(Date,"^2020")) %>% 
  group_by(position) %>% 
  mutate(count = n()) %>% 
  filter(gene == "S") %>% 
  ggplot(fra_merged, mapping = aes(x=position, y=count)) + geom_point() + 
  ggtitle("Spike gene mutations during 2020 in France")



```


```{r}

unique(tidy_300K$pangolin_lineage)


```



Mutations over time for the "5 main lineages"


```{r}



tidy_300K_test <- mutate(tidy_300K, Date = format(substring(tidy_300K$`Collection date`,0,7)))



#Here we use lubridate to make the collection date column into an actual DATE.
tidy_300K_test$`Collection date` <- ymd(tidy_300K_test$`Collection date`) 

  



unique(tidy_300K_test$Variant)



  

tidy_300K_test <- tidy_300K_fra %>% mutate(mutations = (count_N + count_S))

interesting_lineages <- c("B.1.1.7", "P.1", "B.1","B.1.351")

tidy_300K_test <- tidy_300K_test %>% mutate(pango_lineage = ifelse(pangolin_lineage %in% interesting_lineages, pangolin_lineage, "Other lineages"))





tidy_300K_test %>%
  group_by(id, mutations) %>%
  filter(str_detect(Variant, "^VOC")) %>% #This is saying the Variants only starting with "VOC"
  ggplot(tidy_300K_test, mapping = aes(x=`Collection date`,y=mutations, color=Variant)) +
  geom_point() +
  theme(axis.text.x=element_text(angle=90)) +
  theme(legend.title = element_text(size = 3),
         legend.text = element_text(size = 3))+
  ggtitle("Mutations over time for the common lineages")


#Try and remove 2020 2021


?nchar

```


GC content of the 4-5 lineages of interest over time.

```{r}



tidy_300K_test %>%
  group_by(id) %>%
  filter(str_detect(Variant, "^VOC")) %>% #This is saying the Variants only starting with "VOC" 
  ggplot(tidy_300K_test, mapping = aes(x=`Collection date`,y=`GC-Content`, color=Variant)) +
  geom_point(size = 2) +
  geom_smooth(aes(color=Variant), method = "lm", se = F) +
  theme(axis.text.x=element_text(angle=90)) +
  theme(legend.title = element_text(size = 4),
         legend.text = element_text(size = 4)) +
  ggtitle("GC-Content over time for the common lineages") +
  NULL




```



Looking at France only:
```{r}

tidy_300K_test %>% filter(country == "France") %>% 
  filter(str_detect(Variant, "^VOC")) %>% #This is saying the Variants only starting with "VOC" 
  group_by(id) %>%
  ggplot(tidy_300K_test, mapping = aes(x=`Collection date`,y=`GC-Content`, color=Variant)) +
  geom_point() +
  geom_boxplot() +
  theme(axis.text.x=element_text(angle=90)) +
  theme(legend.title = element_text(size = 3),
         legend.text = element_text(size = 3))+
  ggtitle("GC-Content over time for the common lineages in France")+
  facet_wrap( ~ Variant)
  



```



dn/ds for each lineage


```{r}

tidy_300K_test <- tidy_300K_test %>% mutate(DnDs = (count_N/count_S))


tidy_300K_test %>% filter(country == "France") %>% 
  group_by(id) %>%
  ggplot(tidy_300K_test, mapping = aes(x=Date,y=DnDs, color=pango_lineage)) +
  geom_point() +
  theme(axis.text.x=element_text(angle=90)) +
  ggtitle("DnDs ratio for the common lineages in France")+
  facet_wrap( ~ pango_lineage)
  



```


We want to find out whether the mutations are often G/C or otherwise. To do so we look at the reference genome: Wuhan1 and by using a selfwritten function, we use the mutation dataset to see where aswell as which mutation. So 1. we can see which mutations are most prone and also in which regions. We do this because we hypothesize that there are more mutations in the region for immunerecognision associated genes.




```{r}

x <- read_tsv("NC_045512.2.fasta")
reference_genome <- strsplit(x$`>NC_045512.2 Severe acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1, complete genome`[1], "") %>% unlist()
rm(x)
table(reference_genome) %>% knitr::kable()
 

ref_genome <- tibble(base = reference_genome) %>% mutate(position=row_number())




#do it with the frequencies of A C G T?
```



```{r}



bases <- c("T","G","C","A")

dk_merged <- test_merged_df %>% filter(country == "Denmark") %>% filter(variant_base == bases)

#now only variant_base = A T C or G


#Join the wuhan genome to the dk_merged:
  

left_dk <- dk_merged

right_dk <- ref_genome 


dk_merged_with_wuhan <- left_join(left_dk, right_dk, by = "position")


```


```{r}



fahrenheit_to_celsius <- function(temp_F) {
  temp_C <- (temp_F - 32) * 5 / 9
  return(temp_C)
}




fahrenheit_to_celsius(105)


```


```{r}


phdFinder <- function(x){
  for(i in x){
    if(x[i]=="A"){
      res=paste("A")    }
  mentioned <- grepl("A",x)
  return(mentioned)
}
}



x <- c("B","A","C")

phdFinder(dk_merged_with_wuhan$base)





```

